flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/aggfunctions/FirstValueAggFunctionWithRetract.scala

```diff
@@ -0,0 +1,129 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.flink.table.functions.aggfunctions
+
+import java.math.BigDecimal
+import org.apache.flink.api.common.typeinfo.TypeInformation
+import org.apache.flink.table.api.dataview.{ListView, MapView}
+import org.apache.flink.table.functions.AggregateFunction
+
+/** The initial accumulator for first value with retraction aggregate function */
+class FirstValueWithRetractAccumulator[T] {
+  var data: ListView[T] = _
+  var retractedData: MapView[T, Byte] = _
+
+  override def equals(obj: scala.Any): Boolean = {
+    if (obj.isInstanceOf[FirstValueWithRetractAccumulator[T]]) {
+      val target = obj.asInstanceOf[FirstValueWithRetractAccumulator[T]]
+      if (target.data.equals(this.data) && target.retractedData.equals(this.retractedData)) {
+        return true
+      }
+    }
+    false
+  }
+}
+
+/**
+  * Base class for built-in first value with retraction aggregate function
+  *
+  * @tparam T the type for the aggregation result
+  */
+abstract class FirstValueWithRetractAggFunction[T]
+  extends AggregateFunction[T, FirstValueWithRetractAccumulator[T]] {
+
+  override def createAccumulator(): FirstValueWithRetractAccumulator[T] = {
+    val acc = new FirstValueWithRetractAccumulator[T]
+    acc.data = new ListView[T]
+    acc.retractedData = new MapView[T, Byte]
+    acc
+  }
+
+  def accumulate(acc: FirstValueWithRetractAccumulator[T], value: Any): Unit = {
+    if (null != value) {
+      val v = value.asInstanceOf[T]
+      acc.data.add(v)
+    }
+  }
+
+  def retract(acc: FirstValueWithRetractAccumulator[T], value: Any): Unit = {
```

Comment:
It seems that we can do better than having two copies of state here.

It might make sense to extend the ListView interface or to introduce a TreeMap like view.